lib/Transforms/IPO/StructAnalysisTransformAnalyzer.cpp

```diff
@@ -0,0 +1,865 @@
+  if (Field1 == Field2)
+    return 0;
+  auto *Pair = CPBuilder->getCloseProximityPair(Field1, Field2);
+  if (Pair->first < 1e-3)
+    return 0;
+  if (Pair->second < 1)
+    return Pair->first;
+  return Pair->first / Pair->second;
+}
+
+// Calculate weighted CP (WCP) value of all fields within a cache block
+// The function takes an array of fields that can fit on this cache block and
+// returns a sum of CP value between every pair of fields on the block
+double FieldReorderAnalyzer::calculateWCPWithinCacheBlock(
+    std::vector<FieldNumType> *CacheBlock) const {
+  double WCP = 0;
+  DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER,
+                  dbgs() << "Calculate WCP within a cache block: ");
+  for (unsigned i = 0; i < CacheBlock->size(); i++) {
+    DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER,
+                    dbgs() << "F" << (*CacheBlock)[i] + 1 << " ");
+    for (unsigned j = i + 1; j < CacheBlock->size(); j++)
+      if ((*CacheBlock)[i] < (*CacheBlock)[j])
+        WCP += CloseProximityRelations[(*CacheBlock)[i]][(*CacheBlock)[j]];
+      else
+        WCP += CloseProximityRelations[(*CacheBlock)[j]][(*CacheBlock)[i]];
+  }
+  DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER,
+                  dbgs() << "produces WCP: " << DEBUG_PRINT_COUNT(WCP) << "\n");
+  return WCP;
+}
+
+// Calculate a total WCP of all fields that are already in NewOrder. Return
+// the total WCP. This function is used to measure if the current NewOrder
+// is a good ordering
+double FieldReorderAnalyzer::getWCP() const {
+  unsigned CurrentCacheBlockSize = 0;
+  std::vector<FieldNumType> CurrentCacheBlock;
+  double WCP = 0;
+  for (auto &FieldNum : NewOrder) {
+    auto Size = FieldSizes[FieldNum];
+    DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER, dbgs() << "Field " << FieldNum + 1
+                                               << " size is " << Size << "\n");
+    if (!StructureType->getElementType(FieldNum)->isAggregateType() &&
+        CurrentCacheBlockSize % Size) {
+      // Alignment needed, need to add padding. Ignore alignment for aggregated
+      // type (array and struct)
+      DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER,
+                      dbgs() << "Need to add "
+                             << (CurrentCacheBlockSize / Size + 1) * Size -
+                                    CurrentCacheBlockSize
+                             << " Bytes of padding\n");
+      CurrentCacheBlockSize = (CurrentCacheBlockSize / Size + 1) * Size;
+      assert(CurrentCacheBlockSize <= CacheBlockSize);
+    }
+    if (CurrentCacheBlockSize == CacheBlockSize) {
+      DEBUG_WITH_TYPE(
+          DEBUG_TYPE_REORDER,
+          dbgs()
+              << "Need to put into a new cache block calculate current one\n");
+      WCP += calculateWCPWithinCacheBlock(&CurrentCacheBlock);
+      CurrentCacheBlockSize = 0;
+      CurrentCacheBlock.clear();
+    } else {
+      CurrentCacheBlockSize += Size;
+      CurrentCacheBlock.push_back(FieldNum);
+      if (CurrentCacheBlockSize >= CacheBlockSize) {
```

Comment:
When the current field (FieldNum) is pushed to the current cache block, I want to check if it can exceed the CacheBlockSize. If so, I'll only push part of the field to the cache block and calculate the cache block and create a new cache block to hold the remaining part of the field. I'll add a comment to explain it.